pkg/tempdir/tempdir.go

```diff
@@ -0,0 +1,296 @@
+			ids[id] = struct{}{}
+		}
+	}
+	return ids, nil
+}
+
+// RecoverStaleDirs identifies and removes stale temporary directories in the root directory.
+// A directory is considered stale if its lock file can be acquired (indicating no active use).
+// The function attempts to remove both the directory and its lock file.
+// Any errors encountered during the recovery process are collected and returned as a single error.
+// If a directory's lock cannot be acquired, it is considered in use and is skipped.
+func RecoverStaleDirs(rootDir string) error {
+	potentialStaleDirs, err := listPotentialStaleDirs(rootDir)
+	if err != nil {
+		return fmt.Errorf("error listing potential stale temp dirs in %s: %w", rootDir, err)
+	}
+
+	if len(potentialStaleDirs) == 0 {
+		return nil
+	}
+
+	var recoveryErrors []error
+
+	for id := range potentialStaleDirs {
+		lockPath := filepath.Join(rootDir, TempdirLockPrefix+id)
+		tempDirPath := filepath.Join(rootDir, TempDirPrefix+id)
+
+		// Try to lock the lock file. If it can be locked, the directory is stale.
+		instanceLock, err := lockfile.GetLockFile(lockPath)
+		if err != nil {
+			recoveryErrors = append(recoveryErrors, fmt.Errorf("error getting lock file %s: %w", lockPath, err))
+			continue
+		}
+		if err := instanceLock.TryLock(); err != nil {
+			logrus.Debugf("Skipping cleanup for locked directory corresponding to %s", lockPath)
+			continue
+		}
+
+		// Remove the directory and lock file.
+		if rmErr := os.RemoveAll(tempDirPath); rmErr != nil && !os.IsNotExist(rmErr) {
+			recoveryErrors = append(recoveryErrors, fmt.Errorf("error removing stale temp dir %s: %w", tempDirPath, rmErr))
+		}
+
+		defer instanceLock.Unlock()
+		if err := os.Remove(lockPath); err != nil && !os.IsNotExist(err) {
+			recoveryErrors = append(recoveryErrors, fmt.Errorf("error removing stale instance lock file %s: %w", lockPath, err))
+		}
+	}
+
+	return errors.Join(recoveryErrors...)
+}
+
+// initInstanceTempDir creates and locks a new unique temporary directory for this TempDir instance.
+// It creates a temporary lock file with a random name in the root directory,
+// attempts to acquire a lock on it, and then creates the corresponding temp directory.
+// If the lock cannot be acquired after MaxInitRetry attempts, it returns an error.
+// The caller should ensure .Cleanup() is called eventually.
+func (td *TempDir) initInstanceTempDir() error {
+	if td.tempDirPath != "" && td.tempDirLock != nil {
+		logrus.Debugf("Temp dir %s already initialized", td.tempDirPath)
+		return nil
+	}
+
+	// Try to create and lock a unique temporary file
+	try := 0
+	for {
+		// Reserve name for lock file in the RootDir.
+		tempLockFile, err := os.CreateTemp(td.RootDir, TempdirLockPrefix)
+		if err != nil {
+			return fmt.Errorf("creating temp directory in %s failed: %w", td.RootDir, err)
+		}
+		tempLockFile.Close()
+
+		tempDirLockPath := tempLockFile.Name()
+		tempDirLock, err := lockfile.GetLockFile(tempDirLockPath)
+		if err != nil {
```

Comment:
This might fail with `ENOENT` if a concurrent recovery found, locked, and deleted the lock. In that case, retry.